cmd/cosign/cli/fulcio/fulcio.go

```diff
@@ -62,27 +65,54 @@ func (rf *realConnector) OIDConnect(url, clientID, secret, redirectURL string) (
 	return oauthflow.OIDConnect(url, clientID, secret, redirectURL, rf.flow)
 }
 
-func getCertForOauthID(priv *ecdsa.PrivateKey, fc api.LegacyClient, connector oidcConnector, oidcIssuer, oidcClientID, oidcClientSecret, oidcRedirectURL string) (*api.CertificateResponse, error) {
-	pubBytes, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
-	if err != nil {
-		return nil, err
-	}
-
+func getCertForOauthID(sv signature.SignerVerifier, fc api.LegacyClient, connector oidcConnector, oidcIssuer, oidcClientID, oidcClientSecret, oidcRedirectURL string) (*api.CertificateResponse, error) {
 	tok, err := connector.OIDConnect(oidcIssuer, oidcClientID, oidcClientSecret, oidcRedirectURL)
 	if err != nil {
 		return nil, err
 	}
 
-	// Sign the email address as part of the request
 	h := sha256.Sum256([]byte(tok.Subject))
-	proof, err := ecdsa.SignASN1(rand.Reader, priv, h[:])
-	if err != nil {
-		return nil, err
+
+	var pubBytes, proof []byte
+	var algorithm string
+	switch sv := sv.(type) {
+	case *signature.RSAPKCS1v15SignerVerifier:
+		algorithm = "rsa"
+		pubBytes = x509.MarshalPKCS1PublicKey(sv.Public().(*rsa.PublicKey))
+		// Sign the email address as part of the request
+		proof, err = sv.Sign(rand.Reader, h[:], signature.SignerOpts{Hash: crypto.SHA256})
+		if err != nil {
+			return nil, err
+		}
+	case *signature.ECDSASignerVerifier:
+		algorithm = "ecdsa"
+		pubBytes, err = x509.MarshalPKIXPublicKey(sv.Public().(*ecdsa.PublicKey))
+		if err != nil {
+			return nil, err
+		}
+		// Sign the email address as part of the request
+		proof, err = sv.Sign(rand.Reader, h[:], nil)
+		if err != nil {
+			return nil, err
+		}
+	case *signature.ED25519SignerVerifier:
+		algorithm = "ed25519"
+		pubBytes, err = x509.MarshalPKIXPublicKey(sv.Public().(*ed25519.PublicKey))
+		if err != nil {
+			return nil, err
+		}
+		// Sign the email address as part of the request
+		proof, err = sv.Sign(nil, h[:], nil)
```

Comment:
Same here, ed25519 handles computation of the hash, so we shouldn't prehash the message. A test for signing and verifying should cover this